File: src/leetcode/letterCombinationsOfPhoneNumber.cc
// cartesian product 
#include<iostream>
#include<vector>
#include<cstring>
using namespace std;

void output(const vector<string>& vec){
    for(int i = 0; i < vec.size(); i++)
        cout << vec[i] << endl;
}

vector<string> cartesian_product(const string& digits){
    string digit2chars[10] = {"0", "1", "abc",
        "def", "ghi", "jkl", "mno", "opqrs", "tuv", "wxyz"
    };
    if(digits.empty())
        return vector<string>();
    vector<string> res;
    res.push_back("");
    for(int i = 0; i < digits.size(); i++){
        vector<string> temp;
        string chars = digit2chars[digits[i] - '0'];
        for(int j = 0; j < chars.size(); j++){
            for(int k = 0; k< res.size(); k++){
                temp.push_back(res[k] + chars[j]);
            }
        }
        res = temp;
    }
    return res;
}

int main(){
    string digits;
    cin >> digits;
    vector<string> vec = cartesian_product(digits);
    output(vec);
    return 0;
}

File: src/leetcode/letterCombinationsOfPhoneNumber.py
import itertools



if __name__ == "__main__":
    digits = raw_input()
    kvmaps = {"2":"abc","3":"def","4":"ghi","5":"jkl","6":"mno","7":"pqrs","8":"tuv","9":"wxyz"
            }
    #for x in itertools.product(kvmaps[digits[i]] for i in range(len(digits))):
    #for x in itertools.product("abc","def"):
    #    print x
    pools = [kvmaps[digits[i]] for i in range(len(digits))]
    #pools = ["abc","def"]
    '''
    result = [[]]
    for pool in pools:
        result = [x + [y] for x in result for y in pool]
    for prod in result:
        print prod
    
   #for x in itertools.product(*pools):
   '''
    for x in itertools.product(*pools):
       print x
       
